src/datamodel/quests/quest_new.rs

```diff
@@ -8,29 +8,56 @@ pub enum Vote {
     Failed,
 }
 
-pub struct QuestNew<'a> {
-    election: Election<Vote>,
-    winner_rule: &'a (Vote, usize),
+#[derive(Clone)]
+pub struct WinnerRule {
+    pub candidate: Vote,
+    pub electoral_cutoff: usize,
 }
 
-impl<'a> QuestNew<'a> {
-    pub fn new(quest_member: &[String], winner_rule: &'a (Vote, usize)) -> QuestNew<'a> {
-        assert!(quest_member.len() >= winner_rule.1, "Winner's rule cannot be bigger than quest member number");
+pub struct QuestNew {
+    pub election: Election<Vote>,
+    pub winner_rule: WinnerRule,
+}
+
+impl QuestNew {
+    pub fn new(quest_member: &[String], winner_rule: &WinnerRule) -> QuestNew {
+        assert!(
+            quest_member.len() >= winner_rule.electoral_cutoff,
+            "Winner's rule cannot be bigger than quest member number"
+        );
         let election = Election::<Vote>::new(quest_member);
-        QuestNew { election, winner_rule, }
+        QuestNew {
+            election,
+            winner_rule: winner_rule.clone(),
+        }
     }
 
-    pub fn vote(&mut self, quest_member: &String, vote: Vote) {
+    pub fn vote(&mut self, quest_member: &str, vote: Vote) {
         self.election.vote(quest_member, vote);
     }
 
-    pub fn finish_quest(self) -> QuestResult {
-        let scrutiny = self.election.count_votes().ok().unwrap();
-        match (self.winner_rule.0 , scrutiny.result().get(&Some(self.winner_rule.0))) { 
-            (_ , Some(&votes_num)) if votes_num >= self.winner_rule.1 => QuestResult{ quest_result: self.winner_rule.0, },
-            (Vote::Success , _) => QuestResult{ quest_result: Vote::Failed, },
-            (Vote::Failed , _) => QuestResult{ quest_result: Vote::Success, },
-        }
+    pub fn finish_quest(self) -> Result<QuestResult, QuestNew> {
+        let scrutiny = match self.election.count_votes() {
+            Ok(s) => s,
+            Err(e) => {
+                return Err(QuestNew {
+                    election: e,
+                    winner_rule: self.winner_rule,
+                })
+            }
+        };
+        let scrutiny = scrutiny.result();
+        let quest_result = match (
+            self.winner_rule.candidate,
+            scrutiny.get(&Some(self.winner_rule.candidate)),
+        ) {
+            (_, Some(&votes_num)) if votes_num >= self.winner_rule.electoral_cutoff => {
+                self.winner_rule.candidate
+            }
+            (Vote::Success, _) => Vote::Failed,
+            (Vote::Failed, _) => Vote::Success,
+        };
```

Comment:
Ok, if we are leaving this match, then consider using intermediate variables. I would put both elements of the tuple into aux variables which have a name indicating what they are, so the match is more readable by already indicating what it is we are matching:

```rust
        let candidate = self.winner_rule.candidate;
        let election_result = scrutiny.get(&Some(self.winner_rule.candidate);

        let quest_result = match (candidate, election_result) {
            (_, Some(&vote_count)) if vote_count >= self.winner_rule.required_votes => { candidate }
            (Vote::Success, _) => Vote::Failed,
            (Vote::Failed, _) => Vote::Success,
        };
```

Reply:
I'll give it a shot tomorrow but I think second line can fail because of some irrefutably rust rule. You might need to provide a scope to the get to make it work, such as `if let ..... {}`. `match` does it implicitly. I struggle hours on Saturday with that.

Reply:
It's done